File: fidouafclient/src/main/java/io/hanko/fidouafclient/utility/Crypto.java
package io.hanko.fidouafclient.utility;

import android.hardware.fingerprint.FingerprintManager;
import android.os.Build;
import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyProperties;
import android.util.Base64;
import android.util.Log;

import androidx.annotation.Nullable;

import java.io.IOException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.KeyPairGenerator;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.SecureRandom;
import java.security.Signature;
import java.security.SignatureException;
import java.security.UnrecoverableEntryException;
import java.security.cert.CertificateException;
import java.security.spec.ECGenParameterSpec;

public class Crypto {

    private static String KEYSTORE = "AndroidKeyStore";
    private static String TAG = "Crypto";

    public static boolean generateKeyPair(final String keyId) {
        try {
            KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, KEYSTORE);
            KeyGenParameterSpec.Builder keyGenSpecBuilder = new KeyGenParameterSpec.Builder(
                    keyId,
                    KeyProperties.PURPOSE_SIGN
            ).setDigests(KeyProperties.DIGEST_SHA256, KeyProperties.DIGEST_SHA384, KeyProperties.DIGEST_SHA512)
                    .setAlgorithmParameterSpec(new ECGenParameterSpec("secp256r1"))
                    .setUserAuthenticationRequired(true);

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                keyGenSpecBuilder.setInvalidatedByBiometricEnrollment(false);
            }

            keyPairGenerator.initialize(keyGenSpecBuilder.build());
            keyPairGenerator.generateKeyPair();
            return true;
        } catch (NoSuchAlgorithmException | NoSuchProviderException | InvalidAlgorithmParameterException e) {
            Log.e(TAG, "Error while generating KeyPair for fingerprint", e);
            return false;
        }
    }

    public static boolean generateKeyPairForLockscreen(final String keyId) {
        try {
            KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, KEYSTORE);
            keyPairGenerator.initialize(
                    new KeyGenParameterSpec.Builder(
                            keyId,
                            KeyProperties.PURPOSE_SIGN
                    ).setDigests(KeyProperties.DIGEST_SHA256, KeyProperties.DIGEST_SHA384, KeyProperties.DIGEST_SHA512)
                            .setAlgorithmParameterSpec(new ECGenParameterSpec("secp256r1"))
                            .setUserAuthenticationRequired(true)
                            .setUserAuthenticationValidityDurationSeconds(5)
                            .build()
            );

            keyPairGenerator.generateKeyPair();
            return true;
        } catch (NoSuchAlgorithmException | NoSuchProviderException | InvalidAlgorithmParameterException e) {
            Log.e(TAG, "Error while generating KeyPair for lockscreen", e);
            return false;
        }
    }

    @Nullable
    public static String generateKeyID(final String appId) {
        byte[] bytes = new byte[30];
        new SecureRandom().nextBytes(bytes);
        String tmp = appId + Base64.encodeToString(bytes, Base64.DEFAULT);

        try {
            MessageDigest messageDigest = MessageDigest.getInstance("SHA256");
            messageDigest.update(tmp.getBytes());
            return Base64.encodeToString(messageDigest.digest(), Base64.URL_SAFE | Base64.NO_WRAP | Base64.NO_PADDING);
        } catch (NoSuchAlgorithmException e) {
            Log.e(TAG, "Error while generating KeyID", e);
            return null;
        }
    }

    @Nullable
    public static PublicKey getPublicKeyForKeyId(final String keyId) {
        try {
            KeyStore keyStore = KeyStore.getInstance(KEYSTORE);
            keyStore.load(null);
            KeyStore.Entry entry = keyStore.getEntry(keyId, null);
            return ((KeyStore.PrivateKeyEntry) entry).getCertificate().getPublicKey();
        } catch (KeyStoreException | CertificateException | NoSuchAlgorithmException | IOException | UnrecoverableEntryException e) {
            Log.e(TAG, "Error while getting public key", e);
        }
        return null;
    }

    public static void deleteKey(final String keyId) {
        try {
            KeyStore keyStore = KeyStore.getInstance(KEYSTORE);
            keyStore.load(null);
            keyStore.deleteEntry(keyId);
        } catch (KeyStoreException | CertificateException | IOException | NoSuchAlgorithmException e) {
            Log.e(TAG, "Error while deleting key", e);
        }
    }

    @Nullable
    public static byte[] getSignatureForKeyID(final String keyId, final byte[] signedData) {
        try {
            KeyStore keyStore = KeyStore.getInstance(KEYSTORE);
            keyStore.load(null);

            KeyStore.Entry entry = keyStore.getEntry(keyId, null);
            if (!(entry instanceof KeyStore.PrivateKeyEntry)) {
                return null;
            }

            Signature signature = Signature.getInstance("SHA256withECDSA");
            signature.initSign(((KeyStore.PrivateKeyEntry) entry).getPrivateKey());
            signature.update(signedData);
            return signature.sign();
        } catch (KeyStoreException | CertificateException | NoSuchAlgorithmException | UnrecoverableEntryException | IOException | InvalidKeyException | SignatureException e) {
            Log.e(TAG, "Error while getting signature for keyId", e);
        }
        return null;
    }

    public static Signature getSignatureInstance(final String keyId) {
        try {
            KeyStore keyStore = KeyStore.getInstance(KEYSTORE);
            keyStore.load(null);

            PrivateKey privateKey = (PrivateKey) keyStore.getKey(keyId, null);
            Signature signature = Signature.getInstance("SHA256withECDSA");
            signature.initSign(privateKey);
            return signature;
        } catch (KeyStoreException | CertificateException | NoSuchAlgorithmException | UnrecoverableEntryException | IOException | InvalidKeyException e) {
            Log.e(TAG, "Error while getting Signature Instance", e);
        }
        return null;
    }

    public static byte[] getSignature(FingerprintManager.CryptoObject cryptoObject, byte[] signedData) {
        try {
            Signature signature = cryptoObject.getSignature();
            signature.update(signedData);
            return signature.sign();
        } catch (SignatureException e) {
            Log.e(TAG, "Error while getting signature from crypto object", e);
            return null;
        }
    }
}

File: fidouafclient/src/main/java/io/hanko/fidouafclient/util/Crypto.kt
package io.hanko.fidouafclient.util

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import android.util.Log
import java.security.*
import java.security.spec.ECGenParameterSpec

object Crypto {

    const val KEYSTORE = "AndroidKeyStore"
    private const val TAG = "Crypto"

    fun generateKeyPair(keyId: String, appId: String): Boolean {
        try {
            val keyAlias = getKeyStoreAlias(appId, keyId) ?: return false
            val keyPairGenerator = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, KEYSTORE)
            keyPairGenerator.initialize(
                    KeyGenParameterSpec.Builder(
                            keyAlias,
                            KeyProperties.PURPOSE_SIGN
                    ).setDigests(KeyProperties.DIGEST_SHA256, KeyProperties.DIGEST_SHA384, KeyProperties.DIGEST_SHA512)
                            .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
                            .setUserAuthenticationRequired(true)
                            .setUserAuthenticationValidityDurationSeconds(2)
                            .build()
            )

            keyPairGenerator.generateKeyPair()
            return true
        } catch (ex: Exception) {
            Log.e(TAG, "KeyPair could not be generated", ex)
            return false
        }
    }

    fun getStoredKeyIds(appID: String, keyIds: List<String>?): List<String>? {
        val keyStore = KeyStore.getInstance(KEYSTORE).apply { load(null) }
        val storedAliases = keyStore.aliases().toList()

        if (keyIds?.isNotEmpty() == true) {
            val keyAliases = keyIds.map { getKeyStoreAlias(appID, it) }
            return storedAliases.filter { sa -> keyAliases.any { it == sa } }.map { it.split(":")[1] }
        } else {
            val appIdHash = getAppIdHash(appID) ?: return null
            return storedAliases.filter { it.startsWith(appIdHash) }.map { it.split(":")[1] }
        }
    }

    fun generateKeyID(appId: String): String? {
        val bytes = ByteArray(30)
        SecureRandom().nextBytes(bytes)
        val tmp = appId + Base64.encodeToString(bytes, Base64.DEFAULT)

        return try {
            val messageDigest = MessageDigest.getInstance("SHA256")
            messageDigest.update(tmp.toByteArray())
            Base64.encodeToString(messageDigest.digest(), Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
        } catch (ex: NoSuchAlgorithmException) {
            Log.e(TAG, "KeyID could not be generated", ex)
            null
        }
    }

    fun getKeyStoreAlias(appID: String, keyID: String): String? {
        val appIdHash = getAppIdHash(appID)
        return if (appIdHash != null) {
            "$appIdHash:$keyID"
        } else {
            null
        }
    }

    fun getAppIdHash(appID: String): String? {
        return try {
            val messageDigest = MessageDigest.getInstance("SHA256")
            messageDigest.update(appID.toByteArray())
            Base64.encodeToString(messageDigest.digest(), Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
        } catch (ex: NoSuchAlgorithmException) {
            Log.e(TAG, "AppID hash could not be generated", ex)
            null
        }
    }

    fun getPublicKeyForKeyId(keyId: String, appId: String): PublicKey? {
        return try {
            val keyAlias = getKeyStoreAlias(appId, keyId) ?: return null
            val keyStore = KeyStore.getInstance(KEYSTORE).apply { load(null) }

            val storedKeyAlias = keyStore.aliases().toList().find { it.startsWith(keyAlias) }
                    ?: return null

            val entry = keyStore.getEntry(storedKeyAlias, null)
            (entry as KeyStore.PrivateKeyEntry).certificate.publicKey
        } catch (ex: Exception) {
            Log.e(TAG, "Could not get public key for keyId: $keyId and appId: $appId", ex)
            null
        }
    }

    fun deleteKey(keyId: String, appId: String) {
        try {
            val keyAlias = getKeyStoreAlias(appId, keyId) ?: return
            val keyStore = KeyStore.getInstance(KEYSTORE).apply { load(null) }

            val storedKeyAlias = keyStore.aliases().toList().find { it.startsWith(keyAlias) }
                    ?: return

            keyStore.deleteEntry(storedKeyAlias)
        } catch (ex: Exception) {
            Log.e(TAG, "Could not delete key with keyId: $keyId and appId: $appId", ex)
        }
    }

    fun getSignatureForKeyID(keyId: String, signedData: ByteArray, appId: String): ByteArray? {
        try {
            val keyAlias = getKeyStoreAlias(appId, keyId)
            val keyStore = KeyStore.getInstance(KEYSTORE).apply { load(null) }

            val entry = keyStore.getEntry(keyAlias, null)
            if (entry !is KeyStore.PrivateKeyEntry) {
                return null
            }

            val signature = Signature.getInstance("SHA256withECDSA")
            signature.initSign(entry.privateKey)
            signature.update(signedData)
            return signature.sign()
        } catch (ex: Exception) {
            Log.e(TAG, "could not generate signature for keyId: $keyId and appId: $appId", ex)
            return null
        }
    }
}
